python/activator/middleware_interface.py

```diff
@@ -325,8 +325,9 @@ def run_pipeline(self, visit: Visit, snaps: set) -> None:
         where = f"detector={visit.detector} and exposure in ({','.join(str(x) for x in snaps)})"
         pipeline = SimplePipelineExecutor.from_pipeline(self.pipeline, where=where, butler=self.butler)
         _log.info(f"Running pipeline {self.pipeline} on visit '{visit}', snaps {snaps}")
-        # All dataset types should have been pre-registered.
-        result = pipeline.run(register_dataset_types=False)
+        # If this is a fresh (local) repo, then types like calexp,
+        # *Diff_diaSrcTable, etc. have not been registered.
+        result = pipeline.run(register_dataset_types=True)
```

Comment:
Can we do that type registration in prep_butler, possibly as part of `_prep_collections`?

Reply:
That would require a list of every connection used by every task in the pipeline.

Reply:
Doesn't the QuantumGraph know that, though?